beacon-chain/rpc/validator/proposer.go

```diff
@@ -194,6 +206,128 @@ func (vs *Server) eth1Data(ctx context.Context, slot uint64) (*ethpb.Eth1Data, e
+	currentPeriodVotingStartTime := genesisTime +
+		(slot-(slot%slotsPerVotingPeriod))*params.BeaconConfig().SecondsPerSlot
+	previousPeriodVotingStartTime := currentPeriodVotingStartTime -
+		slotsPerVotingPeriod*params.BeaconConfig().SecondsPerSlot
+
+	currentPeriodBlockNumber, err := vs.Eth1BlockFetcher.BlockNumberByTimestamp(ctx, currentPeriodVotingStartTime)
+	if err != nil {
+		log.WithError(err).Error("Failed to get block number for current voting period")
+		return vs.randomETH1DataVote(ctx)
+	}
+	previousPeriodBlockNumber, err := vs.Eth1BlockFetcher.BlockNumberByTimestamp(ctx, previousPeriodVotingStartTime)
+	if err != nil {
+		log.WithError(err).Error("Failed to get block number for previous voting period")
+		return vs.randomETH1DataVote(ctx)
+	}
+
+	headState, err := vs.HeadFetcher.HeadState(ctx)
+	if err != nil {
+		return nil, status.Error(codes.Internal, "Could not get head state")
+	}
+	if len(headState.Eth1DataVotes()) == 0 {
+		eth1Data, err := vs.defaultEth1DataResponse(ctx, currentPeriodBlockNumber)
+		if err != nil {
+			log.WithError(err).Error("Failed to get eth1 data from current period block number")
+			return vs.randomETH1DataVote(ctx)
+		}
+		return eth1Data, nil
+	}
+
+	eth1FollowDistance := int64(params.BeaconConfig().Eth1FollowDistance)
+	lastValidBlockNumber := big.NewInt(0).Sub(currentPeriodBlockNumber, big.NewInt(eth1FollowDistance))
+	firstValidBlockNumber := big.NewInt(0).Sub(previousPeriodBlockNumber, big.NewInt(eth1FollowDistance))
+
+	var inRangeVotes []eth1DataSingleVote
+	for _, eth1Data := range headState.Eth1DataVotes() {
+		ok, height, err := vs.BlockFetcher.BlockExists(ctx, bytesutil.ToBytes32(eth1Data.BlockHash))
+		if err != nil {
+			log.WithError(err).Warning("Could not fetch eth1data height for received eth1data vote")
+		}
+		// TODO: Both inclusive?
+		if ok && lastValidBlockNumber.Cmp(height) > -1 && firstValidBlockNumber.Cmp(height) < 1 {
+			inRangeVotes = append(inRangeVotes, eth1DataSingleVote{eth1Data: *eth1Data, blockHeight: height})
+		}
+	}
+	if len(inRangeVotes) == 0 {
+		eth1Data, err := vs.defaultEth1DataResponse(ctx, currentPeriodBlockNumber)
+		if err != nil {
+			log.WithError(err).Error("Failed to get eth1 data from current period block number")
+			return vs.randomETH1DataVote(ctx)
+		}
+		return eth1Data, nil
+	}
+
+	var voteCount []eth1DataAggregatedVote
+	for _, singleVote := range inRangeVotes {
+		newVote := true
+		for i, aggregatedVote := range voteCount {
+			aggregatedData := aggregatedVote.data
+			if bytes.Equal(singleVote.eth1Data.BlockHash, aggregatedData.eth1Data.BlockHash) &&
+				singleVote.eth1Data.DepositCount == aggregatedData.eth1Data.DepositCount &&
+				bytes.Equal(singleVote.eth1Data.DepositRoot, aggregatedData.eth1Data.DepositRoot) {
+
+				voteCount[i].votes++
+				newVote = false
+				break
+			}
+		}
+
+		if newVote {
+			voteCount = append(voteCount, eth1DataAggregatedVote{data: singleVote, votes: 1})
+		}
```

Comment:
I kind of dislike this algorithm. I wanted to use a simple `map[ethpb.Eth1Data]int` or something to keep track of the votes, but the compiler complained about equality not being defined on the type. I started looking for a solution, but everything was either discouraged from using outside of tests (e.g. `DeepEqual`) or I would need to implement custom equality somehow. All in all I decided to quit looking for now and just implement an obscure and naive algorithm.

Any ideas on how to make this better are welcome!

Reply:
`reflect.DeepEqual`, `ssz.DeepEqual`